File: BackEnd/src/data/mappers/post_mapper.ts
/**
 * @file post_mapper.ts
 * @description Implementación del mapper de posts de la comunidad.
 * Convierte entre PostDocument (MongoDB), Post (dominio) y PostResponseDTO (presentación).
 * @module Community
 * @layer Data
 *
 * @implements {IPostMapper}
 * @injectable
 */

import { injectable } from 'inversify';
import { ObjectId } from 'mongodb';
import type { IPostMapper } from '../IMappers/IPostMapper.js';
import { Post } from '../../domain/entities/Post.js';
import type { PostDocument } from '../datasources/mongodb/models/PostModel.js';
import type { PostResponseDTO } from '../../domain/dtos/community/post-response.dto.js';

/**
 * Mapper de posts de la comunidad.
 *
 * @implements {IPostMapper}
 * @injectable
 */
@injectable()
export class PostMapper implements IPostMapper {

  /**
   * Convierte un documento MongoDB a entidad Post.
   *
   * @param doc — Documento de la colección 'posts'.
   * @returns Entidad Post.
   */
  toEntity(doc: PostDocument): Post {
    return new Post({
      id:            doc._id.toHexString(),
      userId:        doc.userId.toHexString(),
      content:       doc.content,
      image:         doc.image,
      likesCount:    doc.likesCount,
      commentsCount: doc.commentsCount,
      seenBy:        doc.seenBy?.map(id => id.toHexString()) ?? [],
      createdAt:     doc.createdAt,
      updatedAt:     doc.updatedAt,
      deletedAt:     doc.deletedAt,
    });
  }

  /**
   * Convierte una entidad Post (sin id) a documento MongoDB (sin _id).
   *
   * @param entity — Datos del post sin id.
   * @returns Documento parcial para insertar en MongoDB.
   */
  toDocument(entity: Omit<Post, 'id' | 'isActive'>): Omit<PostDocument, '_id'> {
    return {
      userId:        new ObjectId(entity.userId),
      content:       entity.content,
      image:         entity.image,
      likesCount:    entity.likesCount,
      commentsCount: entity.commentsCount,
      seenBy:        entity.seenBy.map(id => new ObjectId(id)),
      createdAt:     entity.createdAt,
      updatedAt:     entity.updatedAt,
      deletedAt:     entity.deletedAt,
    };
  }

  /**
   * Convierte una entidad Post al DTO de respuesta HTTP, enriquecido con datos del autor.
   *
   * @param entity      — Entidad Post.
   * @param authorName  — Nombre del autor del post.
   * @param authorPhoto — URL de la foto del autor (opcional).
   * @param isLikedByMe — true si el usuario que solicita ya dio like (por defecto false).
   * @returns PostResponseDTO serializable.
   */
  toResponseDTO(entity: Post, authorName: string, authorPhoto?: string, isLikedByMe = false): PostResponseDTO {
    return {
      id:            entity.id,
      userId:        entity.userId,
      authorName,
      authorPhoto,
      content:       entity.content,
      image:         entity.image,
      likesCount:    entity.likesCount,
      commentsCount: entity.commentsCount,
      isLikedByMe,
      createdAt:     entity.createdAt.toISOString(),
      updatedAt:     entity.updatedAt.toISOString(),
    };
  }
}

File: FrontEnd/lib/data/mappers/post_mapper.dart
/// @file post_mapper.dart
/// @description Implementación del mapper de posts de comunidad.
/// Convierte PostModel ↔ Post normalizando tipos y parsando fechas.
/// Toda la lógica de transformación vive aquí, nunca en el Model ni en la entidad.
/// @module Community
/// @layer Data
library;

import '../../domain/entities/post.dart';
import '../i_mappers/i_post_mapper.dart';
import '../models/post_model.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// POST MAPPER
// ═══════════════════════════════════════════════════════════════════════════════

/// Implementación de [IPostMapper].
///
/// [implements] IPostMapper
/// [injectable] registrar en container.dart como singleton.
class PostMapper implements IPostMapper {

  // ─── PostModel → Post ─────────────────────────────────────────────────────────

  @override
  Post toEntity(PostModel model) {
    return Post(
      id:            model.id,
      userId:        model.userId,
      authorName:    model.authorName,
      authorPhoto:   model.authorPhoto,
      content:       model.content,
      image:         model.image,
      likesCount:    model.likesCount,
      commentsCount: model.commentsCount,
      isLikedByMe:   model.isLikedByMe,
      // Parsear ISO 8601 → DateTime UTC.
      createdAt:     DateTime.parse(model.createdAt).toUtc(),
      updatedAt:     DateTime.parse(model.updatedAt).toUtc(),
    );
  }

  // ─── Post → PostModel ─────────────────────────────────────────────────────────

  @override
  PostModel toModel(Post entity) {
    return PostModel(
      id:            entity.id,
      userId:        entity.userId,
      authorName:    entity.authorName,
      authorPhoto:   entity.authorPhoto,
      content:       entity.content,
      image:         entity.image,
      likesCount:    entity.likesCount,
      commentsCount: entity.commentsCount,
      isLikedByMe:   entity.isLikedByMe,
      createdAt:     entity.createdAt.toIso8601String(),
      updatedAt:     entity.updatedAt.toIso8601String(),
    );
  }
}
